src/components/common/SpaceTree.js

```diff
@@ -0,0 +1,222 @@
+      {...props}
+    >
+      <path d="M17.485 17.512q-.281.281-.682.281t-.696-.268l-4.12-4.147-4.12 4.147q-.294.268-.696.268t-.682-.281-.281-.682.294-.669l4.12-4.147-4.12-4.147q-.294-.268-.294-.669t.281-.682.682-.281.696 .268l4.12 4.147 4.12-4.147q.294-.268.696-.268t.682.281 .281.669-.294.682l-4.12 4.147 4.12 4.147q.294.268 .294.669t-.281.682zM22.047 22.074v0 0-20.147 0h-20.12v0 20.147 0h20.12zM22.047 24h-20.12q-.803 0-1.365-.562t-.562-1.365v-20.147q0-.776.562-1.351t1.365-.575h20.147q.776 0 1.351.575t.575 1.351v20.147q0 .803-.575 1.365t-1.378.562v0z" />
+    </SvgIcon>
+  );
+};
+
+const SelectedSquare = () => {
+  return <CheckBoxIcon color="primary" />;
+};
+
+const StyledTreeItem = withStyles(theme => ({
+  iconContainer: {
+    '& .close': {
+      opacity: 0.3,
+    },
+  },
+  group: {
+    marginLeft: 7,
+    paddingLeft: 18,
+    borderLeft: `1px dashed ${fade(theme.palette.text.primary, 0.4)}`,
+  },
+  // eslint-disable-next-line react/jsx-props-no-spreading
+}))(props => {
+  const icon = props.selected ? <SelectedSquare /> : null;
+
+  // eslint-disable-next-line react/jsx-props-no-spreading
+  return <TreeItem icon={icon} {...props} />;
+});
+
+const styles = {
+  root: {
+    maxHeight: TREE_VIEW_MAX_HEIGHT,
+    flexGrow: 1,
+    maxWidth: TREE_VIEW_MAX_WIDTH,
+  },
+  verbTitle: {
+    width: '100%',
+  },
+};
+
+class SpaceTree extends Component {
+  static propTypes = {
+    parentSpaceId: PropTypes.string,
+    settings: PropTypes.shape({}).isRequired,
+    dispatchPatchAppInstance: PropTypes.func.isRequired,
+    selectedSpaces: PropTypes.arrayOf(PropTypes.string).isRequired,
+    classes: PropTypes.shape({
+      root: PropTypes.string.isRequired,
+      verbTitle: PropTypes.string.isRequired,
+    }).isRequired,
+    dispatchGetSpaceTree: PropTypes.func.isRequired,
+    tree: PropTypes.arrayOf(PropTypes.shape({}).isRequired).isRequired,
+    expanded: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
+    t: PropTypes.func.isRequired,
+  };
+
+  static defaultProps = {
+    parentSpaceId: null,
+  };
+
+  state = (() => {
+    const { selectedSpaces, parentSpaceId } = this.props;
+    return {
+      selected: selectedSpaces,
+      expanded: [parentSpaceId],
+    };
+  })();
+
+  async componentDidMount() {
+    const { dispatchGetSpaceTree, tree } = this.props;
+
+    // avoid fetching tree each time settings modal is closed
+    if (!tree.length) {
+      dispatchGetSpaceTree();
+    }
+  }
+
+  handleSpaceToggle = newSelectedSpaces => {
+    const { dispatchPatchAppInstance, settings } = this.props;
+    const settingsToChange = {
+      spaces: newSelectedSpaces,
+    };
+    const newSettings = {
+      ...settings,
+      ...settingsToChange,
+    };
+    dispatchPatchAppInstance({
+      data: newSettings,
+    });
+  };
+
+  onNodeSelect = (e, v) => {
+    const { selected } = this.state;
+    const clickedSpaceId = v[0];
+
+    const res = [...selected];
+    const idx = res.indexOf(clickedSpaceId);
```

Comment:
Why do we need to splice / push. Can we not do it using the IDs of the space?

Reply:
Here it handles the select/unselect event, so I either add it when the spaceId is not in the array, and remove it if it exists.